components/camel-smb/src/main/java/org/apache/camel/component/smb/SmbProducer.java

```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    public void disconnect() throws IOException {
+        loggedIn = false;
+        LOGGER.debug("Disconnecting from: {}", getEndpoint());
+
+        session.close();
+
+        session = null;
+    }
+
+    public void createDirectory(DiskShare share, java.io.File file) {
+        String parentDir = file.getParent();
+        SmbConfiguration configuration = getEndpoint().getConfiguration();
+
+        boolean dirExists = share.folderExists(parentDir);
+        if (!dirExists && configuration.isAutoCreate()) {
+            SmbFiles files = new SmbFiles();
+            files.mkdirs(share, normalize(parentDir));
+        }
+
+        if (!dirExists) {
+            throw new RuntimeCamelException("Directory " + parentDir + " does not exist on share " + share.toString());
+        }
+    }
+
+    private GenericFileExist determineFileExist(Exchange exchange) {
+        GenericFileExist gfe = exchange.getIn().getHeader(SmbConstants.SMB_FILE_EXISTS, GenericFileExist.class);
+        if (isEmpty(gfe)) {
+            gfe = getEndpoint().getConfiguration().getFileExist();
+        }
+
+        gfe = (gfe == null) ? GenericFileExist.Fail : gfe;
+
+        return gfe;
+    }
+
+    private void doFail(String path) {
+        throw new RuntimeCamelException("File " + path + " already exists, cannot create");
+    }
+
+    private void doIgnore(String path) {
+        // ignore but indicate that the file was written
+        LOGGER.debug("An existing file already exists: {}. Ignore and do not override it.", path);
+    }
+
+    private void removeFile(DiskShare share, java.io.File file) {
+        LOGGER.debug("An existing file already exists: {}. Removing it.", file.getName());
+        share.rm(file.getPath());
+    }
+
+    @Override
+    public void process(final Exchange exchange) {
+        String fileName = exchange.getIn().getHeader(Exchange.FILE_NAME, String.class);
+
+        SmbConfiguration configuration = getEndpoint().getConfiguration();
+        String path = (configuration.getPath() == null) ? "" : configuration.getPath();
+
+        LOGGER.debug("storeFile() fileName [" + fileName + "]");
+
+        try {
+            connectIfNecessary(exchange);
+
+            java.io.File file = new java.io.File(path + java.io.File.separator + fileName);
+
+            DiskShare share = (DiskShare) session.connectShare(getEndpoint().getShareName());
+            createDirectory(share, file);
+
+            GenericFileExist gfe = determineFileExist(exchange);
+
+            SMB2CreateDisposition scd = SMB2CreateDisposition.FILE_OPEN;
+            HashSet<AccessMask> accessMask = new HashSet<AccessMask>(Arrays.asList(AccessMask.GENERIC_ALL));
+            HashSet<SMB2CreateOptions> createOptions = new HashSet<>(Arrays.asList(SMB2CreateOptions.FILE_DIRECTORY_FILE));
+            HashSet<FileAttributes> fileAttributes
+                    = new HashSet<FileAttributes>(Arrays.asList(FileAttributes.FILE_ATTRIBUTE_NORMAL));
```

Comment:
I would suggest having these as constants: they are in the hot path and we can try to prevent unnecessary overhead.

Reply:
Moved these out of process